core/deployment/src/main/java/io/quarkus/deployment/configuration/BuildTimeConfigurationReader.java

```diff
@@ -512,10 +514,12 @@ ReadResult run() {
                 nameBuilder.setLength(0);
             }
 
+            SmallRyeConfig runtimeConfig = getConfigForRuntimeRecording();
+
             // Register defaults for Roots
-            allBuildTimeValues.putAll(getDefaults(buildTimePatternMap));
-            buildTimeRunTimeValues.putAll(getDefaults(buildTimeRunTimePatternMap));
-            runTimeDefaultValues.putAll(getDefaults(runTimePatternMap));
+            allBuildTimeValues.putAll(getDefaults(config, buildTimePatternMap));
+            buildTimeRunTimeValues.putAll(getDefaults(config, buildTimeRunTimePatternMap));
+            runTimeDefaultValues.putAll(getDefaults(runtimeConfig, runTimePatternMap));
```

Comment:
Just to clarify something: when you start your runtime app, how are the config sources evaluated?

What I'm wondering is if we need to register default values for runtime? Is it because we don't parse the existing config such as `application.properties` at runtime?

Asking because my inclination would have been to record nothing for runtime values and let things be determined at runtime.

Reply:
Correct, we flatten the sources to a single source to optimize lookup and save on loading. We could try to remove the runtime recording and generate the runtime source of each `application.properties` and let it be loaded by the Config.